### PR TITLE
Fix rendering of translucent items in inventory & when wielded

### LINKED ISSUE #5461: Transparent nodes are invisible in inventory
If a node (e.g. MTG water with custom texture) uses texture with alpha less than 128, it’s inventory image is entirely transparent. If the alpha is in range [128, 255], the inventory cube is visible but dark.

### PR DESCRIPTION
**Pull request by appgurueu**
Closes #5461. Ready for review.

## How to test

Look at the three marbles, both in the inventory and as wield items:

* The opaque marble should be opaque.
* The clipped marble should be clipped.
* The blended marble should be blended.

Also look at the alpha test node with alpha = 64. It should be visible, both in your inventory and when wielded. It should look like this:

Screenshot

Cycle through the devtest items and verify that they look right; also look at some simple wield items, like pickaxes, to make sure that there are no regressions.

Furthermore, enable `inventory_items_animations` and check that the spinny items look as expected.

Relevant documentation: https://irrlicht.sourceforge.io/docu/namespaceirr_1_1video.html#ac8e9b6c66f7cebabd1a6d30cbc5430f1ac08aa3715ad41281472202107a81f736

### CODE REVIEW

**Review by sfan5 — APPROVED** [review]
lgtm

**Comment by grorp** [comment]
I've been trying to get through the `wieldmesh.cpp` code and I can't say that I've got everything. However, from what I understand:

In both cases (wield & inventory), the mesh comes from `wieldmesh.cpp`. Wield mesh is from `WieldMeshSceneNode::setItem`, inventory mesh from `getItemMesh` (which share a lot of duplicated code btw). Their child functions already try to set the correct transparency material using `TileLayer::applyMaterialOptions`. On the master branch, this has no chance to work because the transparency material is always overwritten by the user of the mesh.

Now you've made the overwriting work correctly, but the transparency material is still being set somewhere in the mesh generation code too. If possible, I think it would be preferable to make the mesh generation code set the transparency material correctly, getting rid of the overwriting.

**Comment by Wuzzy2** [comment]
Awesome! These broken translucent inventory images for the transparency test in DevTest have bothered me for a long time. Thanks for fixing this.

I found a regression: When you wield `air`, the transparent pixels turn black. If spinning items are enabled, the pixels also turn black when you hover this item in the inventory.

Water sources turn semi-transparent, but the flowing water doesn't. Seems to be only the default flowing water tho. Maybe this is something in the node definition (so not your fault)?

The rest seems fine. I tested wielding all the items (tip: `/hotbar 24` helps a lot), and hovering all the items with spinning items enabled.
I tested wielding the translucent items in front of translucent blocks to find bugs with depth sorting. All looks fine.
Moving the translucent items around in inventory also works fine.

I noticed there is no DevTest test item for testing the case of a `inventory_image` or `wield_image` with translucent pixels. I don't know if Minetest allows this. So I think this test case needs to be added.

**Comment by appgurueu** [comment]
Thanks for the feedback, I'll probably get to addressing it on Thursday.

**Comment by appgurueu** [comment]
> I found a regression: When you wield air, the transparent pixels turn black. If spinning items are enabled, the pixels also turn black when you hover this item in the inventory.

Thanks for finding this. Fixed in https://github.com/minetest/minetest/pull/14701/commits/8c11c12c69fdfd8f001c3ee06aa938bf3f861509:

Screenshot from 2024-06-01 17-29-23

> Seems to be only the default flowing water tho. Maybe this is something in the node definition (so not your fault)?

Indeed this is a devtest bug; the flowing water definition is missing a `..ALPHA`. Fixed in https://github.com/minetest/minetest/pull/14701/commits/f06364e6a5156183aa626d2cb9a8e23f31fe04c4.

**Comment by appgurueu** [comment]
Alright @grorp, I've tried refactoring this (https://github.com/minetest/minetest/pull/14701/commits/ed11b3343053b77d8f43d6b22cf0c8218d71beb5). According to my testing, everything still works. I also tested some additional things:

* Shaders on/off - should be irrelevant
* Translucent liquids on/off - should be consistent
* Dropping items - should look as expected

**Comment by Wuzzy2** [comment]
Now every item glow in the wieldhand and as dropped item. Test the wield image at night/darkness and you'll see.

**Comment by appgurueu** [comment]
> Now every item glow in the wieldhand and as dropped item. Test the wield image at night/darkness and you'll see.

I'm not sure I can reproduce. This is after `/time 0` in devtest:

Screenshot from 2024-06-02 14-35-41

vs. `/time 12000`:

Screenshot from 2024-06-02 14-35-48

Tested with https://github.com/minetest/minetest/pull/14701/commits/ed11b3343053b77d8f43d6b22cf0c8218d71beb5. I also reviewed the code - I shouldn't haven't touched any `Lighting = true/false` assignments for materials.

**Comment by grorp** [comment]
Can reproduce with `enable_shaders = true`, cannot reproduce with `enable_shaders = false`. See also #9525 and related PRs.

The `Lighting` material flag is only for the fixed pipeline IIRC.

**Comment by grorp** [comment]
Posting this thing here too:

# The mighty `wieldmesh.cpp`

https://github.com/minetest/minetest/pull/14701 as of commit https://github.com/minetest/minetest/pull/14701/commits/ed11b3343053b77d8f43d6b22cf0c8218d71beb5

## `WieldMeshSceneNode::setItem`

determines material at the very beginning via `ContentFeatures::getMaterialType` and `MaterialType_to_irr`, saves it as `m_material_type`/`m_material_param`

cases:

- `WieldMeshSceneNode::setExtruded`
    - respects `m_material_type`/`m_material_param`
- `WieldMeshSceneNode::setCube`
    - uses `TileLayer::applyMaterialOptions` *and* respects `m_material_type`/`m_material_param` (???)
- `createSpecialNodeMesh`
    - uses `TileLayer::applyMaterialOptions`

=> diagnosis: inconsistent

## `getItemMesh`

cases:

- `getExtrudedMesh`
    - always uses `EMT_TRANSPARENT_ALPHA_CHANNEL`/`0.0f`, material is overwritten later
    - *mostly duplicate of `WieldMeshSceneNode::setExtruded`*
- `createCube` + `postProcessNodeMesh`
    - doesn't set any material, material is overwritten later
    - *maybe duplicate of `WieldMeshSceneNode::setCube`?* (but doesn't use `TileLayer::applyMaterialOptions`)
- `createSpecialNodeMesh`
    - uses `TileLayer::applyMaterialOptions` *and* material is overwritten later (???)

sets material at the very end via `ContentFeatures::getMaterialType` and `MaterialType_to_irr`, by overwriting all materials of the already created mesh

=> diagnosis: inconsistent

## Open questions

- What about `TileLayer::applyMaterialOptions`? only two usages, inconsistent
- What about `TileLayer::applyMaterialOptionsWithShaders`?
- There was code for rendering wieldmeshes using shader materials before, why is it gone?

**Comment by appgurueu** [comment]
`WieldMeshSceneNode::setCube`: Got rid of the `m_material_type` / `m_material_type_param` usage there, thanks. That should make `setItem` consistent: `applyMaterialOptions` is used for nodes, `m_material_type[_param]` for extrusion meshes (items, nodes that use inventory image / wield image).

`getItemMesh` is a mess. I'll try to clean it up. Note that for `getExtrudedMesh`, the material is not overwritten later on all paths.

**Comment by grorp** [comment]
> I still find it weird that we first give our materials to our shader to then get an Irrlicht material type (?) back, rather than having our shader deal properly with the material type it is given.

I believe it returns a "shader material" with an index past the builtin materials and that's how the shader is applied.

**Comment by appgurueu** [comment]
> I believe it returns a "shader material" with an index past the builtin materials and that's how the shader is applied.

Yes. This is quite a dirty hack IMO, but I won't fix it here. Looking at the enum, this is unexpected (though `EMT_FORCE_32BIT` should probably have raised my eyebrow).

Something like `std::variant<E_MATERIAL_TYPE, ShaderMaterialId>` would probably make more sense.

---

Anyways, I think as far as this fix goes, this PR should be okay now? Please check whether I've missed anything.

File: src/client/wieldmesh.cpp [hunk 1]
```diff
@@ -41,6 +41,56 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #define MIN_EXTRUSION_MESH_RESOLUTION 16
 #define MAX_EXTRUSION_MESH_RESOLUTION 512
 
+/*!
```
Comment (appgurueu) [inline src/client/wieldmesh.cpp:44]: (I just moved this around because I made it `static`.)

File: src/client/wieldmesh.cpp [hunk 2]
```diff
@@ -193,10 +243,55 @@ class ExtrusionMeshCache: public IReferenceCounted
 
 static ExtrusionMeshCache *g_extrusion_mesh_cache = nullptr;
 
+static scene::SMesh *getExtrudedMesh(ITextureSource *tsrc,
```
Comment (appgurueu) [inline src/client/wieldmesh.cpp:246]: (Same here)

**Comment by sfan5** [comment]
> I believe it returns a "shader material" with an index past the builtin materials and that's how the shader is applied.

Indeed.
Irrlicht has some built-in materials that are implemented either using shaders or by the fixed pipeline.
When you register a pair of vertex / fragment shaders that itself becomes a new "material" to use and it shares the enumeration space with the built-in ones.

> Something like `std::variant<E_MATERIAL_TYPE, ShaderMaterialId>` would probably make more sense.

This would have unneeded overhead and make few sense given internal handling is often identical.

File: src/client/wieldmesh.cpp [hunk 3]
```diff
@@ -365,16 +469,21 @@ void WieldMeshSceneNode::setItem(const ItemStack &item, Client *client, bool che
...
 
+	const auto material_type = def.type == ITEM_NODE
+			? f.getMaterialType() : TILE_MATERIAL_ALPHA;
+	MaterialType_to_irr(material_type, m_material_type, m_material_type_param);
 	if (m_enable_shaders) {
-		u32 shader_id = shdrsrc->getShader("object_shader", TILE_MATERIAL_BASIC, NDT_NORMAL);
+		IShaderSource *shdrsrc = client->getShaderSource();
+		u32 shader_id = shdrsrc->getShader("object_shader",
+				material_type, def.type == ITEM_NODE ? f.drawtype : NDT_MESH);
+		// Note: The shader may give us a different material back.
```
Comment (sfan5) [inline src/client/wieldmesh.cpp:486]: it will, not may

**Comment by grorp** [comment]
<details>
<summary>New summary</summary>
<h2 id="-wieldmeshscenenode-setitem-"><code>WieldMeshSceneNode::setItem</code></h2>
<p>at the very beginning, <code>m_material_type</code>/<code>m_material_type_param</code> are set via <code>ContentFeatures::getMaterialType</code>/<code>MaterialType_to_irr</code>/<code>getShader</code></p>
<p>cases:</p>
<ul>
<li><code>WieldMeshSceneNode::setExtruded</code> (<code>!wield_image.empty() &amp;&amp; check_wield_image</code> branch, fallback else branch)<ul>
<li>material set from <code>m_material_type</code>/<code>m_material_type_param</code></li>
</ul>
</li>
<li><code>WieldMeshSceneNode::setExtruded</code> (<code>def.type == ITEM_NODE</code> branch)<ul>
<li>material set from <code>m_material_type</code>/<code>m_material_type_param</code></li>
<li>material overriden with <code>m_material_type</code>/<code>m_material_type_param</code> later</li>
</ul>
</li>
<li><code>WieldMeshSceneNode::setCube</code> (<code>def.type == ITEM_NODE</code> branch)<ul>
<li>material set via <code>TileLayer::applyMaterialOptions</code> (probably doesn&#39;t take shaders into account?)</li>
<li>material overriden with <code>m_material_type</code>/<code>m_material_type_param</code> later</li>
</ul>
</li>
<li><code>createSpecialNodeMesh</code> (<code>def.type == ITEM_NODE</code> branch)<ul>
<li>material not set</li>
<li>material overriden with <code>m_material_type</code>/<code>m_material_type_param</code> later</li>
</ul>
</li>
</ul>
<h2 id="getitemmesh">getItemMesh</h2>
<p>cases:</p>
<ul>
<li><code>getExtrudedMesh</code> (<code>!inventory_image.empty()</code> branch, <code>def.type == ITEM_NODE &amp;&amp; f.drawtype == NDT_AIRLIKE</code> branch)<ul>
<li>material always set to <code>EMT_TRANSPARENT_ALPHA_CHANNEL</code>/<code>0.0f</code></li>
</ul>
</li>
<li><code>getExtrudedMesh</code> (<code>def.type == ITEM_NODE</code> branch)<ul>
<li>material always set to <code>EMT_TRANSPARENT_ALPHA_CHANNEL</code>/<code>0.0f</code></li>
<li>material overriden with <code>TileLayer::applyMaterialOptions</code> later</li>
</ul>
</li>
<li><code>ExtrusionMeshCache::createCube</code> (<code>def.type == ITEM_NODE</code> branch)<ul>
<li>material not set</li>
<li>material overriden with <code>TileLayer::applyMaterialOptions</code> later</li>
</ul>
</li>
<li><code>createSpecialNodeMesh</code> (<code>def.type == ITEM_NODE</code> branch)<ul>
<li>material not set</li>
<li>material overriden with <code>TileLayer::applyMaterialOptions</code> later</li>
</ul>
</li>
</ul>
</details>

Anyway, this PR causes a regression with the texture overlay test nodes:

master branch
master branch

this PR
this PR

**Pull request closed without merging**